flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -337,6 +338,9 @@ public final void invoke() throws Exception {
 			operatorChain = new OperatorChain<>(this, recordWriters);
 			headOperator = operatorChain.getHeadOperator();
 
+			// check environment for selective reading
+			checkSelectiveReadingEnv();
```

Comment:
If validated on the JM side, one way to determine whether a job contains selective reading operator is to mark it by adding a boolean field in `JobGraph` when generating JobGraph, and another way is to deserialize `StreamOperatorFactory` from `StreamConfig`. I don't think the second way is good, because JM currently does not need to deserialize `StreamConfig`, which will increase the time cost. Considering that the default value of `taskmanager.network.credit-model` is true, this check is a protective code and it does not normally trigger a check exception. In addition, `taskmanager.network.credit-model` is an option that is deprecated, and the validation will not be needed after it is removed. So I think it's easier to add validation here and to remove it in the future.

Reply:
My main suspect was that you may not have the same environment configuration at JM's side.

But regarding `OperatorChain.hasSelectiveReadingOperator()` implementation, I think delegating it to JM should not add more overhead. You already have
```java
Class<?> operatorClass = operatorFactory.getStreamOperatorClass();
InputSelectable.class.isAssignableFrom(operatorClass);
```
check in [`StreamingJobGraphGenerator`](https://github.com/apache/flink/blob/9ab0b7d3b23c8c548edd0723be34e143e144afaf/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java#L682-L683).

Reply:
I'm not sure if what you're talking about is the first way I said. If yes, I don't think it's necessary to add a boolean field to the job graph for the checking logic which will be removed in the future. If not, I understand you're talking about the second way as I said, that `StreamOperatorFactory` needs to be deserialized from `StreamConfig` on the JM side, which will increase the time cost ( Please note that JM currently does not need to deserialize `StreamConfig`). If neither method is not, I don't know what your proposal is like.

Reply:
My understanding was that `StreamingJobGraphGenerator.createJobGraph()` happens in JM side only. But it looks like it is also called at sql-client, where you may not have the same environment configuration.

> taskmanager.network.credit-model is an option that is deprecated, and the validation will not be needed after it is removed. So I think it's easier to add validation here and to remove it in the future.

I'm fine with this.